ctapipe/image/muon/muon_processor.py

```diff
@@ -68,71 +55,48 @@ class MuonAnalysis(Tool):
         help="Pedestal noise rms", default_value=1.1
     ).tag(config=True)
 
-    classes = [
-        CameraCalibrator,
-        TailcutsImageCleaner,
-        EventSource,
-        MuonRingFitter,
-        MuonIntensityFitter,
-    ]
-
-    aliases = {
-        "i": "EventSource.input_url",
-        "input": "EventSource.input_url",
-        "o": "MuonAnalysis.output",
-        "output": "MuonAnalysis.output",
-        "max-events": "EventSource.max_events",
-        "allowed-tels": "EventSource.allowed_tels",
-    }
-
-    flags = {
-        "overwrite": ({"MuonAnalysis": {"overwrite": True}}, "overwrite output file")
-    }
-
-    def setup(self):
-        if self.output is None:
-            raise ToolConfigurationError("You need to provide an --output file")
-
-        if self.output.exists() and not self.overwrite:
-            raise ToolConfigurationError(
-                "Outputfile {self.output} already exists, use `--overwrite` to overwrite"
-            )
-
-        self.source = EventSource(parent=self)
-        subarray = self.source.subarray
-
-        self.calib = CameraCalibrator(subarray=subarray, parent=self)
+    def __init__(
+        self, subarray: SubarrayDescription, config=None, parent=None, **kwargs
+    ):
+        """
+        Parameters
+        ----------
+        subarray: SubarrayDescription
+            Description of the subarray. Provides information about the
+            camera which are useful in calibration. Also required for
+            configuring the TelescopeParameter traitlets.
+        config: traitlets.loader.Config
+            Configuration specified by config file or cmdline arguments.
+            Used to set traitlet values.
+            This is mutually exclusive with passing a ``parent``.
+        parent: ctapipe.core.Component or ctapipe.core.Tool
+            Parent of this component in the configuration hierarchy,
+            this is mutually exclusive with passing ``config``
+        """
+
+        super().__init__(subarray=subarray, config=config, parent=parent, **kwargs)
+
+        self.calib = CameraCalibrator(subarray=self.subarray, parent=self)
```

Comment:
This will either overwrite the calibration or simply not run, depending on the order.  Calibration should not be part of this class - it should assume the calibration is already done in the outer event loop (which it is already), since it is schematically like this:

```py3
for event in source:
     calibrate(event)
     process_images(event)
     process_muons(event)
```

Reply:
No, this is intended. We want to be able to use different calibration settings for the Muon events.

This will overwrite the images, but only after those have been written to the output file.

Reply:
that sounds like something that will come back to bite us- if two calibrations (one hidden inside another object) overwrites the internal event structure, it means you cannot run this in any order you want, and you may end up with images that are not what you expect.  I really don't like that.

Reply:
Would it not be better to just  re-do the image extraction part internally, rather than run an entire CameraCalibrator?  The idea of the calibration steps was that they were modular and could be re-used inside other steps of the analysis

Reply:
The main point why we wanted to integrate the muon analysis here was to avoid opening the files and doing the low-level calibrations multiple times. As @moralejo pointed out, we need to use a different extraction method for the muons.

If the problem is changing the images in the structure, we can deepcopy the events in the muon processor.

Reply:
But you *are* are doing low-level analysis multiple times,.   I do not want a DL1 data model that contains images and parameters, but the parameters are not computed from the images, but rather from some other hidden version of those image.  That is a recipe for confusion. 

I think we should have a call about this - it will not be resolved in the issue tracker.

Reply:
> I still really don't like the idea that there are parameters computed from images that are not saved, however.

The same happens by default in stage1, parameters are stored but not images.

> Would it not be better to just re-do the image extraction part internally, rather than run an entire CameraCalibrator? 

That would mean also copying all the corrections the that happen inside the `CameraCalibrator`. I really don't see a benefit.

CameraCalibrator does only two things, DVR in R1→ DL0, and image extraction in DL0 → DL1. Sure, we only need the second part, but right now that's not possible, that would mean splitting the DVR and DL0→DL1 part into two components.

Reply:
> But you are are doing low-level analysis multiple times,.

I am talking here about the LST R0 → R1 corrections, which are the most computationally expensive, not about the image extraction.

Reply:
I think there are only two possible solutions, we include it here in ctapipe, or lstchain will need to do it in their own version of the stage1 tool.

Which would also already be a big improvement over the current situation, but not what we really want.

Reply:
I agree with @kosack that storing images obtained with a pulse integrator which is not the same with which image parameters  are produced is very misleading. But what is the problem with the deepcopy solution proposed by @maxnoe?  
Muon ring events would have their standard image parameters, from the standard image extraction. Besides, they would have the muon ring parameters obtained with the GlobalPeak pulse integrator. 

Eventually, one might consider to store, for muon rings, _also_ the images extracted with the GlobalPeak in a separate table. 

Is there any technical problem with the above approach?

Reply:
I thought about this a bit, and there is a simple solution (and sorry to @RuneDominik  that I think it is similar to what he started with!). 

The answer is that we are really talking about two DL1 files:  one for the regular analysis + parameters as we do now, and one for the muon charge extraction + parameters.   So we should just write both separately.  That way both are correct DL1 files and the "provenance chain" of image → parameters is consistent in each.  That solves also the problem that the length of the muon parameters dataset would not have be the same as the parameter dataset, as you would only write out events for which the muon detection routine thinks there is a muon.  We have an implicit assumption in the DL1 files now that we write parameters for *all* written images, and this way we can keep that assumption. 

So just add an option to set the muon output file (unless you want to auto name it by pre-pending "muon" or something), and add a second `DL1DataWriter` called `muon_writer` where `muon_writer.write()`  is only called for muon-like events.  You can also just overwrite the DL1 container twice as you do now, but we should mention this very explicitly as a caveat in the docs so that if we decide to  change the order of components, parallelize, or something unexpected, it doesn't come back to surprise us.  Ideally do the muon version first, call `muon_writer.write()then overwrite it with the normal DL1 info. 

In the end you will just get two files, as if you ran stage1 twice, but without the extra io overhead, and although this is a bit of a hack, it could just be removed in future versions once we have explicit muon-selected data files.